File: p5js/js/animsprite.js
"use strict";

class AnimSprite { // _name is single PImage

    constructor(_name, _fps, _tdx, _tdy, _etx, _ety) {
        this.spriteName = "sprite";
        this.spriteTag = "tag";
        this.framesArray = []; // PImage array
        this.loopIn = 0;
        this.currentFrame = parseFloat(this.loopIn);
        this.loopOut = 1; // this.framesArray.length; 
        this.fps = _fps;
        this.rate = 0.2; // 12/60
        this.playing = true;
        this.playOnce = false;
        this.debug = false;

        this.mouseable = true;
        this.hovered = false;
        this.clicked = false;

        //position, rotation, scale, target
        this.r = 0;
        this.p = createVector(0, 0);
        this.s = createVector(0, 0);
        this.t = createVector(0, 0);
        this.w = 10;
        this.h = 10;
        this.ease = 100;
        this.shake = 5;
        this.hitSpread = 5;

        this.loadSpriteSheet(_name, _tdx,_tdy,_etx,_ety);
        this.init(_tdx, _tdy);

        this.gravity = 1.0;
        this.floor = 400;

        console.log(this.spriteName + " :: " + 
                                "frames: " + this.framesArray.length +
                                "    loop in: " + this.loopIn +
                                "    loop out: " + this.loopOut + 
                                "    fps: " + this.fps);
    }

    //~~~~~~~~~~~~~~~~~~~~~~
    init(_tdx, _tdy) {
        this.p = createVector(0, 0);
        this.r = 0;
        this.s = createVector(1, 1);
        this.t = createVector(0, 0);
        
        this.w = _tdx;//this.framesArray[0].width;
        this.h = _tdy;//this.framesArray[0].height;    
    }

    loadSpriteSheet(_name, _tdx, _tdy, _etx, _ety) {
        try {
            //loads a spritesheet from a single image
            var tileX = 1;
            var tileY = 1;
            var tileDimX = _tdx;
            var tileDimY = _tdy;
            var endTileX = _etx;
            var endTileY = _ety;

            for (var i = 0; i < _etx*_ety; i++) {
                if (tileX + tileDimX <= (endTileX * tileDimX)) {
                    tileX += tileDimX;
                } else if (tileY + tileDimY <= (endTileY * tileDimY)) {
                    tileY += tileDimY;
                    tileX = 1;
                } else {
                    tileX = 1;
                    tileY = 1; 
                }
                //println("Loading frame" + (h+1) + " from " + _name);
                this.framesArray.push(_name.get(tileX, tileY, tileDimX, tileDimY));
            }
            this.loopOut = this.framesArray.length;

        } catch(e) { 
            console.log("Image loading failed.");
        }
    }

    //~~~~~~~~~~~~~~~~~~~~~~

    setRate() {
        this.rate = this.fps/frameRate();
    }

    update() {
        if (this.mouseable) {
            this.checkHover();
            this.checkClick();    
        }

        this.setRate();
        if(this.playing) {
            this.currentFrame += this.rate;
            if (this.currentFrame >= this.loopOut) {
                if(this.playOnce) {
                    this.playing=false;
                } else {
                    this.currentFrame = this.loopIn;
                }
            }
        }
    }

    draw() {
        pg.push();
        //translate, rotate, scale
        pg.translate(this.p.x, this.p.y);
        pg.rotate(radians(this.r));
        pg.scale(this.s.x, this.s.y);

        //draw
        pg.imageMode(CENTER);
        pg.image(this.framesArray[parseInt(this.currentFrame)], 0, 0);
        pg.pop();
    }

    run() {
        this.update();
        this.draw();
    }

    //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    //basic behaviors

    checkHover() {
        this.hovered = this.hitDetect(mouseX,mouseY);
    }

    checkClick() {
        if (this.hovered) this.clicked = mouseIsPressed;
    }

    //Tween movement.    start, end, ease (more = slower).
    tween(_e) { //float
        var _ease;
        if (!_e) {
            _ease = this.ease;
        } else {
            _ease = _e;
        }
        this.p.x += (this.t.x-this.p.x)/_ease;
        this.p.y += (this.t.y-this.p.y)/_ease;
    }

    shaker(_s) {
        var _shake;
        if (!_s) {
            _shake = this.shake;
        } else {
            _shake = _s;
        }    
        this.p.x += random(_shake) - random(_shake);
        this.p.y += random(_shake) - random(_shake);
    }

    bounds() {
        if (this.p.x < 0) {
            this.p.x = 0;
        } else if (this.p.x > pg.width) {
            this.p.x = pg.width;
        }
        if (this.p.y < 0) {
            this.p.y = 0;
        } else if (this.p.y > pg.height) {
            this.p.y = pg.height;
        }
    }

    falling(_g) {    //y pos, floor num, gravity num
        var _gravity;
        if (!_g) {
            _gravity = this.gravity;
        } else {
            _gravity = _g;
        }

        if (this.p.y < this.floor) {
            this.p.y += _gravity;
        } else if (this.p.y >= this.floor) {
            this.p.y = this.floor;
        }
    }

    //2D Hit Detect.    Assumes center.    x,y,w,h of object 1, x,y,w,h, of object 2.
    hitDetect(_x, _y) {
        var x1 = this.p.x;
        var y1 = this.p.y;
        var w1 = this.w/2;
        var h1 = this.h/2;
        var x2;
        if (!_x) {
            x2 = this.t.x;
        } else {
            x2 = _x;
        }    
        var y2;
        if (!_y) {
            y2 = this.t.y;
        } else {
            y2 = _y;
        } 
        var w2 = this.hitSpread/2;
        var h2 = this.hitSpread/2; 
        if (x1 + w1 >= x2 - w2 && x1 - w1 <= x2 + w2 && y1 + h1 >= y2 - h2 && y1 - h1 <= y2 + h2) {
            return true;
        } else {
            return false;
        }
    }

}

File: processingpy/AnimSpriteExample/AnimSprite.py
# uses bits from...
class AnimSprite:
    frames = []
    loopIn = 0
    index = loopIn
    loopOut, fps
    speed
    playing = true
    playOnce = false
    is3D = False
    isTexture = False
    debug = False
    hovered = False
    clicked = False
    #position, rotation, scale, target
    r
    p,r3D,s,t
    uvs = []
    vertices = []
    vertices_proj = []
    w
    h
    ease = 100

    #folder of frames method
    AnimSprite(_name, _fps):
        loadFrames(_name)
        loopOut = frames.length 
        fps = _fps
        init()

    AnimSprite(PImage[] _name, int _fps):
        frames = _name
        loopOut = frames.length 
        fps = _fps
        init()
    
    #spritesheet method
    AnimSprite(String _name, int _fps, int _tdx, int _tdy, int _etx, int _ety):
        loadSpriteSheet(_name, _tdx,_tdy,_etx,_ety)
        loopOut = frames.length 
        fps = _fps
        init()

    #~~~~~~~~~~~~~~~~~~~~~~
    def init():
        if (is3D == False):
            p = new PVector(0, 0)
            r = 0
            s = new PVector(1, 1)
            t = new PVector(0, 0)
        else:
            p = new PVector(0, 0, 0)
            r3D = new PVector(0, 0, 0)
            s = new PVector(1, 1)
            t = new PVector(0, 0, 0)
        
        w = frames[0].width
        h = frames[0].height
    
    def make3D():
        is3D = True
        isTexture = False
        init()
    }
    
    def makeTexture():
        is3D = true
        isTexture = true
        init()
        vertices[0] = new PVector(-1*(frames[0].width/2),-1*(frames[0].height/2))
        vertices[1] = new PVector(frames[0].width/2,-1*(frames[0].height/2))
        vertices[2] = new PVector(frames[0].width/2,frames[0].height/2)
        vertices[3] = new PVector(-1*(frames[0].width/2),frames[0].height/2)
        uvs[0] = new PVector(0,0)
        uvs[1] = new PVector(frames[0].width,0)
        uvs[2] = new PVector(frames[0].width,frames[0].height)
        uvs[3] = new PVector(0,frames[0].height) 
    
    def make2D():
        is3D = False
        isTexture = False
        init()
    
    vertToProj(_p, _centerPoint):
        return new PVector(_p.x + _centerPoint.x,_p.y + _centerPoint.y)
    
    projToVert(_p, _centerPoint):
        return new PVector(_p.x - _centerPoint.x,_p.y - _centerPoint.y)
    
    #~~~~~~~~~~~~~~~~~~~~~~
    
    def loadFrames(_name):
        try:
            #loads a sequence of frames from a folder
            filesCounter=0
            File dataFolder = new File(dataFile(sketchPath()).getAbsolutePath(), "data/"+_name) 
            allFiles = dataFolder.list()
            for i in range(0, len(allFiles)):
                if (allFiles[i].toLowerCase().endsWith("png")):
                    filesCounter++
            #--
            frames = []
            for i in range(0, filesCounter):
                println("Loading " + _name + "/frame" + (i+1) + ".png")
                frames[i] = loadImage("data/"+_name + "/frame" + (i+1) + ".png")
        catch(Exception e):
            pass
    
    def loadSpriteSheet(_name, _tdx, _tdy, _etx, _ety):
            try:
                #loads a spritesheet from a single image
                fromImg = loadImage("data/"+_name + ".png")
                tileX = 1
                tileY = 1
                tileDimX = _tdx
                tileDimY = _tdy
                endTileX = _etx
                endTileY = _ety
                #--
                frames = []
                for i in range(0, _etx*_ety):
                    if tileX + tileDimX <= (endTileX*tileDimX):
                        tileX += tileDimX
                    elif tileY + tileDimY<=(endTileY*tileDimY):
                        tileY += tileDimY
                        tileX = 1
                    else:
                        tileX = 1
                        tileY = 1

                    println("Loading frame" + (h+1) + " from " + _name + ".png")
                    frames[h] = fromImg.get(tileX, tileY, tileDimX, tileDimY)
            catch(Exception e):
                pass

#~~~~~~~~~~~~~~~~~~~~~~

    def setSpeed(_fps):
        speed = _fps/parseFloat(frameRate)
    
    def update():
        setSpeed(fps)
        if(playing):
            index += speed
            if (index >= loopOut):
                if (playOnce):
                    playing = False
                else:
                    index = loopIn

    def draw():
        frameIndex = parseInt(index)
        pushMatrix()
        #translate, rotate, scale
        if (is3D == False)
            translate(p.x, p.y)
            rotate(radians(r))
            scale(s.x, s.y)
        else:
            translate(p.x, p.y, p.z)
            rotate(radians(r))
            rotateXYZ(r3D.x, r3D.y, r3D.z)
            scale(s.x, s.y)
        
        #draw
        if (isTexture == False):
            imageMode(CENTER)
            image(frames[frameIndex], 0, 0)
        else:
            for i in range(0, len(vertices)):
                vertices_proj[i] = vertToProj(vertices[i],p)
                
            noFill()
            noStroke()
            beginShape(QUADS)
            texture(frames[frameIndex])
            vertex(vertices[0].x,vertices[0].y,uvs[0].x,uvs[0].y)
            vertex(vertices[1].x,vertices[1].y,uvs[1].x,uvs[1].y)
            vertex(vertices[2].x,vertices[2].y,uvs[2].x,uvs[2].y)
            vertex(vertices[3].x,vertices[3].y,uvs[3].x,uvs[3].y)
            endShape(CLOSE)
            
            if(debug){
                noStroke()
                ellipseMode(CENTER)
                fill(255,0,0)
                ellipse(vertices[0].x,vertices[0].y,5,5)
                fill(0,255,0)
                ellipse(vertices[1].x,vertices[1].y,5,5)
                fill(0,0,255)
                ellipse(vertices[2].x,vertices[2].y,5,5)
                fill(255,0,255)
                ellipse(vertices[3].x,vertices[3].y,5,5)
        popMatrix()
        
    def run():
        update()
        draw()
        
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    #utilities

    def rotateXYZ(_x, _y, _z):
        rotateX(radians(_x))
        rotateY(radians(_y))
        rotateZ(radians(_z))

    #simplifies the unnecessarily complex blend command image, x, y, width, height, center/corner
    def blendImage(PImage bI, int pX, int pY, String b, boolean center) {
        String[] blendModes = { 
            "BLEND", "ADD", "SUBTRACT", "LIGHTEST", "DARKEST", "DIFFERENCE", "EXCLUSION", "MULTIPLY", "SCREEN", "OVERLAY", "HARD_LIGHT", "SOFT_LIGHT", "DODGE", "BURN"
        }
        int[] blendModeCodes = { 
            1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192
        }
        for (int i=0i<blendModes.lengthi++) {
            if (b==blendModes[i]) {
                int qX, qY
                if (center) {
                    qX=pX-(bI.width/2)
                    qY=pY-(bI.height/2)
                } 
                else {
                    qX=pX
                    qY=pY
                }
                blend(bI, 0, 0, bI.width, bI.height, qX, qY, bI.width, bI.height, blendModeCodes[i])
            }
        }
    }

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    #basic behaviors

    #Tween movement.    start, end, ease (more = slower).
    float tween(float v1, float v2, float e) {
        v1 += (v2-v1)/e
        return v1
    }

    PVector tween3D(PVector v1, PVector v2, PVector e) {
        v1.x += (v2.x-v1.x)/e.x
        v1.y += (v2.y-v1.y)/e.y
        v1.z += (v2.z-v1.z)/e.z
        return v1
    }
    
    float shake(float v1, float s) {
        v1 += random(s) - random(s)
        return v1
    }

    float boundary(float v1, float vMin, float vMax) {
        if (v1<vMin) {
            v1 = vMin
        } 
        else if (v1>vMax) {
            v1=vMax
        } 
        return v1
    }

    float gravity(float v1, float v2, float v3) { #y pos, floor num, gravity num
        if (v1<v2) {
            v1 += v3
        }
        if (v1>v2) {
            v1 = v2
        }
        return v1
    }

    #2D Hit Detect.    Assumes center.    x,y,w,h of object 1, x,y,w,h, of object 2.
    boolean hitDetect(float x1, float y1, float w1, float h1, float x2, float y2, float w2, float h2) {
        w1 /= 2
        h1 /= 2
        w2 /= 2
        h2 /= 2 
        if (x1 + w1 >= x2 - w2 && x1 - w1 <= x2 + w2 && y1 + h1 >= y2 - h2 && y1 - h1 <= y2 + h2) {
            return true
        } 
        else {
            return false
        }
    }
    
        #3D Hit Detect.    Assumes center.    xyz, whd of object 1, xyz, whd of object 2.
   def hitDetect3D(p1, s1, p2, s2):
        s1.x /= 2
        s1.y /= 2
        s1.z /= 2
        s2.x /= 2
        s2.y /= 2 
        s2.z /= 2 
        if (    p1.x + s1.x >= p2.x - s2.x && 
                    p1.x - s1.x <= p2.x + s2.x && 
                    p1.y + s1.y >= p2.y - s2.y && 
                    p1.y - s1.y <= p2.y + s2.y &&
                    p1.z + s1.z >= p2.z - s2.z && 
                    p1.z - s1.z <= p2.z + s2.z
            )
            return True
        else:
            return False